src/atomics/acquire-release.md

```diff
@@ -0,0 +1,334 @@
+            └────────┘           │
+            ┌────────┬───────────┘
+            │ false  │
+            └────────┘
+```
+
+Look at the second operation Thread 2 performs (the read of `data`), for which
+we haven’t yet joined the line. Where should it connect to? Well actually, it
+has multiple options…wait, we’ve seen this before! It’s a data race!
+
+That’s not good. Last time the solution was to use atomics instead — but in this
+case that doesn’t seem to be enough, since even if atomics were used it still
+would have the _option_ of reading `0` instead of `1`, and really if we want our
+mutex to be sane, it should only be able to read `1`.
+
+So it seems that want we _want_ is to be able to apply our arrow rules from
+before to completely rule out zero from the set of the possible values — if we
+were able to draw a large arrow from the Thread 1’s `+= 1;` to Thread 2’s
+`guard`, then we could trivially then use the rule to rule out `0` as a value
+that could be read.
+
+This is where the `Acquire` and `Release` orderings come in. Informally put, a
+_release store_ will cause an arrow instead of a line to be drawn from the
+operation to the destination; and similarly an _acquire load_ will cause an
+arrow to be drawn from the destination to the operation. To give a useless
+example that illustrates this, for the given program:
+
+```rust
+# use std::sync::atomic::{self, AtomicU32};
+// Initial state
+let a = AtomicU32::new(0);
+// Thread 1
+a.store(1, atomic::Ordering::Release);
+// Thread 2
+a.load(atomic::Ordering::Acquire);
+```
+
+The two possible executions look like this:
+
+```text
+    Possible Execution 1      ┃      Possible Execution 2
+                              ┃
+Thread 1      a     Thread 2  ┃  Thread 1      a     Thread 2
+╭───────╮   ┌───┐   ╭──────╮  ┃  ╭───────╮   ┌───┐   ╭──────╮
+│ store ├─┐ │ 0 │ ┌─→ load │  ┃  │ store ├─┐ │ 0 ├───→ load │
+╰───────╯ │ └───┘ │ ╰──────╯  ┃  ╰───────╯ │ └───┘   ╰──────╯
+          └─↘───┐ │           ┃            └─↘───┐
+            │ 1 ├─┘           ┃              │ 1 │
+            └───┘             ┃              └───┘
+```
+
+These arrows are a new kind of arrow we
+haven’t seen yet; they are known as _happens before_ (or happens after)
+relations and are represented as thin arrows (→) on these diagrams. They are
+weaker than the _sequenced before_ double-arrows (⇒) that occur inside a single
+thread, but can still be used with the arrow rules to determine which values of
+a memory location are valid to read. We can say that in the first possible
+execution, Thread 1’s `store` (and everything sequenced before that) _happens
+before_ Thread 2’s load (and everything sequenced after that).
+
+There is one more rule required for these to be useful, and that is _release
+sequences_: after a release store is performed on an atomic, happens before
+arrows will connect together each subsequent value of the atomic as long as the
+new value is caused by an RMW and not just a plain store.
```

Comment:
Don't release stores participate in the release sequence as well or is this a gap in my understanding?

Reply:
Until C++20, stores performed by the original thread participated in release sequences but I don’t think arbitrary release stores ever have. [cppreference says this](https://en.cppreference.com/w/cpp/atomic/memory_order#Release_sequence):

> After a release operation A is performed on an atomic object M, the longest continuous subsequence of the modification order of M that consists of
> 1) Writes performed by the same thread that performed A (until C++20)
> 2) Atomic read-modify-write operations made to M by any thread
> is known as _release sequence headed by A_

Reply:
Interesting. Might be worth calling that caveat out explicitly then.